cuda_pathfinder/cuda/pathfinder/_headers/find_nvidia_headers.py

```diff
@@ -4,39 +4,114 @@
+    return None
+
 
-    if libname == "nvshmem" and IS_WINDOWS:
+def _joined_isfile(dirpath: str, basename: str) -> bool:
+    return os.path.isfile(os.path.join(dirpath, basename))
+
+
+def _find_nvshmem_header_directory() -> Optional[str]:
+    if IS_WINDOWS:
         # nvshmem has no Windows support.
         return None
 
     # Installed from a wheel
     nvidia_sub_dirs = ("nvidia", "nvshmem", "include")
     hdr_dir: str  # help mypy
     for hdr_dir in find_sub_dirs_all_sitepackages(nvidia_sub_dirs):
-        nvshmem_h_path = os.path.join(hdr_dir, "nvshmem.h")
-        if os.path.isfile(nvshmem_h_path):
+        if _joined_isfile(hdr_dir, "nvshmem.h"):
             return hdr_dir
 
     conda_prefix = os.environ.get("CONDA_PREFIX")
     if conda_prefix and os.path.isdir(conda_prefix):
         hdr_dir = os.path.join(conda_prefix, "include")
-        nvshmem_h_path = os.path.join(hdr_dir, "nvshmem.h")
-        if os.path.isfile(nvshmem_h_path):
+        if _joined_isfile(hdr_dir, "nvshmem.h"):
             return hdr_dir
 
     for hdr_dir in sorted(glob.glob("/usr/include/nvshmem_*"), reverse=True):
-        nvshmem_h_path = os.path.join(hdr_dir, "nvshmem.h")
-        if os.path.isfile(nvshmem_h_path):
+        if _joined_isfile(hdr_dir, "nvshmem.h"):
             return hdr_dir
 
     return None
+
+
+def _find_based_on_ctk_layout(libname: str, h_basename: str, anchor_point: str) -> Optional[str]:
+    if libname == "nvvm":
+        idir = os.path.join(anchor_point, "nvvm", "include")
+        if _joined_isfile(idir, h_basename):
+            return idir
+    else:
+        idir = os.path.join(anchor_point, "include")
+        if libname == "cccl":
+            cdir = os.path.join(idir, "cccl")
+            if _joined_isfile(cdir, h_basename):
+                return cdir
+        if _joined_isfile(idir, h_basename):
+            return idir
+
+    return None
+
+
+@functools.cache
+def _get_nvcc_home() -> Optional[str]:
+    # Fall back to the directory that owns nvcc (works for most local installs)
+    nvcc_path = shutil.which("nvcc")
+    if not nvcc_path:
+        return None
+    flds = nvcc_path.split(os.path.sep)
+    if len(flds) < 3:
+        return None
+    return os.path.sep.join(flds[:-2])
+
+
+def _find_ctk_header_directory(libname: str) -> Optional[str]:
+    h_basename = supported_nvidia_headers.SUPPORTED_HEADERS_CTK[libname]
+    candidate_dirs = supported_nvidia_headers.SUPPORTED_SITE_PACKAGE_HEADER_DIRS_CTK[libname]
+
+    # Installed from a wheel
+    for cdir in candidate_dirs:
+        hdr_dir: str  # help mypy
+        for hdr_dir in find_sub_dirs_all_sitepackages(tuple(cdir.split("/"))):
+            if _joined_isfile(hdr_dir, h_basename):
+                return hdr_dir
+
+    conda_prefix = os.getenv("CONDA_PREFIX")
+    if conda_prefix:  # noqa: SIM102
+        if result := _find_based_on_ctk_layout(libname, h_basename, conda_prefix):
+            return result
```

Comment:
This is what we were trying to say in an offline thread. This is strictly speaking unsafe/wrong depending on `libname`. 
- In general, CUDA headers are installed to `$CONDA_PREFIX/targets/<target_name>/include`, including `cuda-cccl` (the conda package, not the wheel!) 
   - `<target_name>` can be found [here](https://github.com/conda-forge/cuda-cudart-feedstock/blob/3266304793943a796088c7216c7801a97645de5a/recipe/meta.yaml#L9-L12)
- `cccl` (another conda package) lives in `$CONDA_PREFIX/include`
- non-CTK libs are installed to `$CONDA_PREFIX/include` too (I am 99% sure)

Reply:
Sorry I should have left a comment: this was just meant to be a placeholder/starting point. I will work on this tomorrow (manual testing under Linux and Windows).